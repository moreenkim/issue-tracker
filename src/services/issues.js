import axios from 'axios';

class IssueService {
  authenticateUser() {
    return axios.create({
      baseURL: process.env.REACT_APP_GITHUB_GRAPHQL_API_BASE_URL,
      headers: {
        Authorization: `Bearer ${process.env.REACT_APP_GITHUB_ACESS_TOKEN}`,
      },
    });
  }

  async getRepoIssues({ path, state, pageNum, filters }) {
    // Return empty array when path is an empty string. Triggered by the search clear button.
    if (path === '') {
      return [];
    }

    const [owner, repoName] = path.split('/');
    const query = `
      query($owner: String!, $repoName: String!, $state: IssueState!, $pageNum: Int, $issueFilters: IssueFilters) {
        repository(owner: $owner, name: $repoName) {
          issues(first: $pageNum, states:[$state], filterBy: $issueFilters) {
            edges {
              node {
                id
                title
                url
                labels(first:5) {
                  edges {
                    node {
                      name
                    }
                  }
                }
              }
            }
          }
        }
    }
    `;

    const response = await this.authenticateUser().post('', {
      query,
      variables: {
        owner,
        repoName,
        state: state || 'OPEN',
        pageNum: pageNum || 20,
        issueFilters: filters || {},
      },
    });

    // return empty array if no repo found
    if (response.data.data.repository === null) {
      return [];
    }

    return response.data.data.repository.issues.edges;
  }

  async getRepoLabels({ path }) {
    // Return empty array when path is an empty string. Triggered by the search clear button.
    if (path === '') {
      return [];
    }

    const [owner, repoName] = path.split('/');
    const query = `
      query($owner: String!, $repoName: String!) {
        repository(owner: $owner, name: $repoName) {
          labels(first: 100) {
            edges {
              node {
                name
              }
            }
          }
        }
    }
    `;

    const response = await this.authenticateUser().post('', {
      query,
      variables: {
        owner,
        repoName,
      },
    });

    // return empty array if no repo found
    if (response.data.data.repository === null) {
      return [];
    }

    return response.data.data.repository.labels.edges;
  }
}

export default new IssueService();
