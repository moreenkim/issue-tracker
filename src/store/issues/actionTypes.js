export const GET_ISSUES = 'issues.GET_ISSUES';
export const GET_ISSUE = 'issues.GET_ISSUE';
