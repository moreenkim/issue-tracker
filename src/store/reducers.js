import issues from './issues/reducer';

export { issues };
