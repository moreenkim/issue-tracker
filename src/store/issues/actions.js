import IssueService from '../../services/issues';
import * as types from './actionTypes';

export function getRepoIssuesAction(payload) {
  return async (dispatch) => {
    try {
      const issues = await IssueService.getRepoIssues(payload);

      dispatch({
        type: types.GET_ISSUES,
        issues,
      });
    } catch (error) {
      console.error(error);
    }
  };
}
