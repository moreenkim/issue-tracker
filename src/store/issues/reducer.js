import Immutable from 'seamless-immutable';

import * as types from './actionTypes';

const initialState = Immutable({
  issues: null,
  issue: {},
});

export default function reduce(state = initialState, action = {}) {
  switch (action.type) {
    case types.GET_ISSUES:
      return state.merge({
        issues: action.issues,
      });
    case types.GET_ISSUE:
      return state.merge({
        issue: action.issue,
      });
    default:
      return state;
  }
}

//selectors, get data from store
export function getIssues(state) {
  return state.issues.issues;
}

export function getIssue(state) {
  return state.issues.issue;
}
