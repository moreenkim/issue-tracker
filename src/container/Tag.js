import React, { Component } from 'react';
import { Tag, Col } from 'antd';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as issuesSelectors from '../store/issues/reducer';
import * as issuesActions from '../store/issues/actions';

class RepoTags extends Component {
  constructor(props) {
    super(props);
    this.state = {
      path: props.path,
    };
  }

  async componentDidMount() {
    const { getRepoLabels, labels } = this.props;
    const { path } = this.state;
    // only get repository labels when labels array is empty
    if (labels.asMutable().length === 0) {
      await getRepoLabels({ path });
    }
  }

  async handleChange(label, checked) {
    const {
      issueState,
      path,
      getRepoIssues,
      setSelectedLabels,
      selectedLabels,
    } = this.props;
    const nextselectedLabels = checked
      ? [...selectedLabels, label.node.name]
      : selectedLabels.filter((t) => t !== label.node.name);
    setSelectedLabels(nextselectedLabels);
    // get issues by repo name and issue state
    await getRepoIssues({
      path,
      state: issueState,
      filters: nextselectedLabels.length ? { labels: nextselectedLabels } : {},
    });
  }

  render() {
    const { CheckableTag } = Tag;
    const { labels, selectedLabels } = this.props;
    return (
      <>
        <Col>
          <span style={{ fontSize: '15px' }}>Issue Labels:</span>
        </Col>
        <Col>
          {labels.asMutable().length > 0 ? (
            labels.map((label) => (
              <CheckableTag
                style={{ fontSize: '15px' }}
                key={label.node.name}
                checked={selectedLabels.indexOf(label.node.name) > -1}
                onChange={(checked) => this.handleChange(label, checked)}
              >
                {label.node.name}
              </CheckableTag>
            ))
          ) : (
            <span>No Labels</span>
          )}
        </Col>
      </>
    );
  }
}

function mapStateToProps(state, props) {
  return {
    labels: issuesSelectors.getLabels(state),
    selectedLabels: issuesSelectors.getSelectedLabels(state),
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      getRepoLabels: issuesActions.getRepoLabelsAction,
      getRepoIssues: issuesActions.getRepoIssuesAction,
      setSelectedLabels: issuesActions.setSelectedLabelsAction,
    },
    dispatch
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(RepoTags);
