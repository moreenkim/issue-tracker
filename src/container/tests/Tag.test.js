// Test Tag
import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { BrowserRouter } from 'react-router-dom';
import { createStore, applyMiddleware, combineReducers } from 'redux';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import * as reducers from '../../store/reducers';

import * as issuesActionTypes from '../../store/issues/actionTypes';

import Tag from '../Tag';

describe('containers/Tag', () => {
  let store;

  beforeEach(() => {
    store = createStore(combineReducers(reducers), applyMiddleware(thunk));
  });

  it('renders without crashing', () => {
    shallow(
      <Provider store={store}>
        <BrowserRouter>
          <Tag path="owner/repo-name" issueState="OPEN" />
        </BrowserRouter>
      </Provider>
    );
  });

  it('renders tags correctly', () => {
    store.dispatch({
      type: issuesActionTypes.GET_LABELS,
      labels: [
        {
          node: {
            name: 'bugs',
          },
        },
      ],
    });
    const wrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <Tag path="owner/repo-name" issueState="OPEN" />
        </BrowserRouter>
      </Provider>
    );
    expect(wrapper.text()).toContain('bugs');
    expect(wrapper.text()).toContain('Issue Labels:');
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
