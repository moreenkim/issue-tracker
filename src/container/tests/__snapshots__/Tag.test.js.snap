// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Tag renders tags correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(RepoTags)
        issueState="OPEN"
        path="owner/repo-name"
      >
        <RepoTags
          getRepoIssues={[Function]}
          getRepoLabels={[Function]}
          issueState="OPEN"
          labels={
            Array [
              Object {
                "node": Object {
                  "name": "bugs",
                },
              },
            ]
          }
          path="owner/repo-name"
          selectedLabels={Array []}
          setSelectedLabels={[Function]}
        >
          <Col>
            <div
              className="ant-col"
              style={Object {}}
            >
              <span
                style={
                  Object {
                    "fontSize": "15px",
                  }
                }
              >
                Issue Labels:
              </span>
            </div>
          </Col>
          <Col>
            <div
              className="ant-col"
              style={Object {}}
            >
              <CheckableTag
                checked={false}
                key="bugs"
                onChange={[Function]}
                style={
                  Object {
                    "fontSize": "15px",
                  }
                }
              >
                <span
                  className="ant-tag ant-tag-checkable"
                  onClick={[Function]}
                  style={
                    Object {
                      "fontSize": "15px",
                    }
                  }
                >
                  bugs
                </span>
              </CheckableTag>
            </div>
          </Col>
        </RepoTags>
      </Connect(RepoTags)>
    </Router>
  </BrowserRouter>
</Provider>
`;
